File: tr-4/blockchain.py
"""
Prototipo de DLT (Blockchain)
- Operación append para agregar bloques
- Prueba de trabajo simple (Proof of Work) con dificultad de 4 ceros
- Sistema de marca de tiempo (timestamp)
"""

import hashlib
import time
import json


DIFICULTAD = 4  # El hash debe comenzar con 4 ceros: "0000..."
PREFIJO_POW = "0" * DIFICULTAD


class Bloque:
    """Representa un bloque individual en la cadena."""

    def __init__(self, indice: int, datos, hash_previo: str):
        self.indice = indice
        self.marca_tiempo = time.time()  # timestamp UNIX
        self.datos = datos
        self.hash_previo = hash_previo
        self.nonce = 0
        self.hash = self._minar()

    # ------------------------------------------------------------------
    # Cálculo del hash
    # ------------------------------------------------------------------

    def _calcular_hash(self) -> str:
        """Devuelve el SHA-256 del bloque con el nonce actual."""
        contenido = json.dumps({
            "indice": self.indice,
            "marca_tiempo": self.marca_tiempo,
            "datos": self.datos,
            "hash_previo": self.hash_previo,
            "nonce": self.nonce,
        }, sort_keys=True)
        return hashlib.sha256(contenido.encode()).hexdigest()

    # ------------------------------------------------------------------
    # Prueba de trabajo (Proof of Work)
    # ------------------------------------------------------------------

    def _minar(self) -> str:
        """Incrementa el nonce hasta que el hash comience con PREFIJO_POW."""
        print(f"  Minando bloque #{self.indice}...", end=" ", flush=True)
        inicio = time.time()

        hash_candidato = self._calcular_hash()
        while not hash_candidato.startswith(PREFIJO_POW):
            self.nonce += 1
            hash_candidato = self._calcular_hash()

        duracion = time.time() - inicio
        print(f"listo en {duracion:.3f}s  (nonce={self.nonce})")
        return hash_candidato

    # ------------------------------------------------------------------
    # Representación
    # ------------------------------------------------------------------

    def marca_tiempo_legible(self) -> str:
        """Convierte el timestamp UNIX a formato ISO 8601."""
        return time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(self.marca_tiempo))

    def __repr__(self) -> str:
        return (
            f"Bloque #{self.indice}\n"
            f"  Marca de tiempo : {self.marca_tiempo_legible()}\n"
            f"  Datos           : {self.datos}\n"
            f"  Nonce           : {self.nonce}\n"
            f"  Hash previo     : {self.hash_previo}\n"
            f"  Hash            : {self.hash}\n"
        )


# ----------------------------------------------------------------------

class Blockchain:
    """Cadena de bloques con operación append y validación."""

    def __init__(self):
        self._cadena: list[Bloque] = []
        print("=== Creando bloque génesis ===")
        self._cadena.append(self._crear_genesis())

    # ------------------------------------------------------------------
    # Bloque génesis
    # ------------------------------------------------------------------

    def _crear_genesis(self) -> Bloque:
        """El primer bloque usa hash previo de ceros."""
        return Bloque(
            indice=0,
            datos="Bloque Génesis",
            hash_previo="0" * 64,
        )

    # ------------------------------------------------------------------
    # Operación append
    # ------------------------------------------------------------------

    def append(self, datos) -> Bloque:
        """Agrega un nuevo bloque al final de la cadena."""
        ultimo = self._cadena[-1]
        nuevo = Bloque(
            indice=len(self._cadena),
            datos=datos,
            hash_previo=ultimo.hash,
        )
        self._cadena.append(nuevo)
        return nuevo

    # ------------------------------------------------------------------
    # Validación de la cadena
    # ------------------------------------------------------------------

    def es_valida(self) -> bool:
        """
        Verifica la integridad de toda la cadena:
        1. Cada hash debe cumplir la prueba de trabajo.
        2. Cada hash_previo debe coincidir con el hash del bloque anterior.
        """
        for i in range(1, len(self._cadena)):
            actual = self._cadena[i]
            anterior = self._cadena[i - 1]

            # Re-calcula el hash esperado
            contenido = json.dumps({
                "indice": actual.indice,
                "marca_tiempo": actual.marca_tiempo,
                "datos": actual.datos,
                "hash_previo": actual.hash_previo,
                "nonce": actual.nonce,
            }, sort_keys=True)
            hash_esperado = hashlib.sha256(contenido.encode()).hexdigest()

            if actual.hash != hash_esperado:
                print(f"  ✗ Hash inválido en bloque #{actual.indice}")
                return False
            if not actual.hash.startswith(PREFIJO_POW):
                print(f"  ✗ Prueba de trabajo fallida en bloque #{actual.indice}")
                return False
            if actual.hash_previo != anterior.hash:
                print(f"  ✗ Eslabón roto entre bloque #{i-1} y #{i}")
                return False

        return True

    # ------------------------------------------------------------------
    # Acceso e información
    # ------------------------------------------------------------------

    def __len__(self) -> int:
        return len(self._cadena)

    def __getitem__(self, indice: int) -> Bloque:
        return self._cadena[indice]

    def __iter__(self):
        return iter(self._cadena)

    def imprimir_cadena(self):
        """Muestra todos los bloques de la cadena."""
        print("\n" + "=" * 60)
        print(f"  CADENA DE BLOQUES  ({len(self._cadena)} bloque(s))")
        print("=" * 60)
        for bloque in self._cadena:
            print(bloque)

File: tr-4/src/blockchain.rs
//! Prototipo de DLT (Blockchain)
//! - Operación append
//! - Prueba de trabajo (Proof of Work) con dificultad de 4 ceros
//! - Sistema de marca de tiempo (timestamp)

use chrono::{DateTime, Local, TimeZone};
use sha2::{Digest, Sha256};
use std::time::{Instant, SystemTime, UNIX_EPOCH};

const DIFICULTAD: usize = 4;
const PREFIJO_POW: &str = "0000";

// ─────────────────────────────────────────────────────────────────────────────
// Bloque
// ─────────────────────────────────────────────────────────────────────────────

pub struct Bloque {
    pub indice: u64,
    pub marca_tiempo: u64,   // Unix timestamp en segundos
    pub datos: String,
    pub hash_previo: String,
    pub nonce: u64,
    pub hash: String,
}

impl Bloque {
    /// Crea un nuevo bloque y ejecuta la prueba de trabajo.
    pub fn nuevo(indice: u64, datos: String, hash_previo: String) -> Self {
        let marca_tiempo = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .expect("Error de reloj del sistema")
            .as_secs();

        let mut bloque = Bloque {
            indice,
            marca_tiempo,
            datos,
            hash_previo,
            nonce: 0,
            hash: String::new(),
        };

        bloque.hash = bloque.minar();
        bloque
    }

    // ── Cálculo del hash ────────────────────────────────────────────────────

    pub fn calcular_hash(&self) -> String {
        // Serialización ordenada (mismo orden siempre → hash determinista)
        let contenido = format!(
            r#"{{"datos":{},"hash_previo":"{}","indice":{},"marca_tiempo":{},"nonce":{}}}"#,
            serde_json::to_string(&self.datos).unwrap(),
            self.hash_previo,
            self.indice,
            self.marca_tiempo,
            self.nonce,
        );
        let resultado = Sha256::digest(contenido.as_bytes());
        format!("{:x}", resultado)
    }

    // ── Prueba de trabajo (Proof of Work) ───────────────────────────────────

    fn minar(&mut self) -> String {
        print!("  Minando bloque #{}... ", self.indice);
        let inicio = Instant::now();

        loop {
            let hash = self.calcular_hash();
            if hash.starts_with(PREFIJO_POW) {
                println!(
                    "listo en {:.3}s  (nonce={})",
                    inicio.elapsed().as_secs_f64(),
                    self.nonce
                );
                return hash;
            }
            self.nonce += 1;
        }
    }

    // ── Marca de tiempo legible ─────────────────────────────────────────────

    pub fn marca_tiempo_legible(&self) -> String {
        let dt: DateTime<Local> = Local
            .timestamp_opt(self.marca_tiempo as i64, 0)
            .single()
            .unwrap_or_else(|| Local::now());
        dt.format("%Y-%m-%d %H:%M:%S").to_string()
    }
}

impl std::fmt::Display for Bloque {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Bloque #{}\n  Marca de tiempo : {}\n  Datos           : {}\n  Nonce           : {}\n  Hash previo     : {}\n  Hash            : {}\n",
            self.indice,
            self.marca_tiempo_legible(),
            self.datos,
            self.nonce,
            self.hash_previo,
            self.hash,
        )
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Blockchain
// ─────────────────────────────────────────────────────────────────────────────

pub struct Blockchain {
    cadena: Vec<Bloque>,
}

impl Blockchain {
    /// Crea la cadena y genera el bloque génesis.
    pub fn nueva() -> Self {
        println!("=== Creando bloque génesis ===");
        let genesis = Bloque::nuevo(0, "Bloque Génesis".to_string(), "0".repeat(64));
        Blockchain { cadena: vec![genesis] }
    }

    // ── Operación append ────────────────────────────────────────────────────

    /// Agrega un nuevo bloque al final de la cadena.
    pub fn append(&mut self, datos: String) -> &Bloque {
        let hash_previo = self.cadena.last().unwrap().hash.clone();
        let indice = self.cadena.len() as u64;
        let bloque = Bloque::nuevo(indice, datos, hash_previo);
        self.cadena.push(bloque);
        self.cadena.last().unwrap()
    }

    // ── Validación ──────────────────────────────────────────────────────────

    /// Verifica integridad completa de la cadena.
    pub fn es_valida(&self) -> bool {
        for i in 1..self.cadena.len() {
            let actual = &self.cadena[i];
            let anterior = &self.cadena[i - 1];

            let hash_esperado = actual.calcular_hash();

            if actual.hash != hash_esperado {
                println!("  ✗ Hash inválido en bloque #{}", actual.indice);
                return false;
            }
            if !actual.hash.starts_with(PREFIJO_POW) {
                println!("  ✗ Prueba de trabajo fallida en bloque #{}", actual.indice);
                return false;
            }
            if actual.hash_previo != anterior.hash {
                println!(
                    "  ✗ Eslabón roto entre bloque #{} y #{}",
                    i - 1,
                    i
                );
                return false;
            }
        }
        true
    }

    // ── Acceso ──────────────────────────────────────────────────────────────

    pub fn len(&self) -> usize {
        self.cadena.len()
    }

    pub fn get_mut(&mut self, indice: usize) -> Option<&mut Bloque> {
        self.cadena.get_mut(indice)
    }

    pub fn imprimir_cadena(&self) {
        let linea = "=".repeat(60);
        println!("\n{}", linea);
        println!("  CADENA DE BLOQUES  ({} bloque(s))", self.cadena.len());
        println!("{}", linea);
        for bloque in &self.cadena {
            println!("{}", bloque);
        }
    }
}
